addOns/pscanrulesBeta/src/main/java/org/zaproxy/zap/extension/pscanrulesBeta/InformationDisclosureSuspiciousComments.java

```diff
@@ -125,24 +128,29 @@ private void raiseAlert(HttpMessage msg, int id, String detail) {
         parent.raiseAlert(id, alert);
     }
 
-    private List<Pattern> getPatterns() {
-        if (patterns == null) {
-            patterns = new ArrayList<>();
-            String line = null;
-            File f = new File(Constant.getZapHome() + File.separator + databaseErrorFile);
```

Comment:
The file was being loaded from the file system to allow the user to customise the entries to its own needs.

Reply:
This is a key feature of these scan rules - if the user cant easily change them then they are much less useful.

Reply:
OK. So how do you set up a unit test so when the scanner is loaded to test, it actually loads the default suspicious-comments.txt file properly?  When running the unit test, without making this change, it can't find that file so the scanner can't find anything. So we changed it to be a resource so it can find it in the unit test, and presumably when ZAP is running on its own. If there is a way to make it work for both junit tests and within ZAP, and be replaceable by the user when desired, please explain how, so it can be changed to do that.

Reply:
You can change the scanner to allow to specify a test file (better as it doesn't rely on the production file), `InformationDisclosureDebugErrorsUnitTest` does that. If you prefer to use the same file you can copy it into the (test) ZAP home (given by `Constant.getZapHome()`).

Reply:
> The file was being loaded from the file system to allow the user to customise the entries to its own needs.

How would a user change that file in the current implementation when using ZAP (before this suggested change)? The changes proposed don't change where the file is, it just changes how it is loaded so the unit test can successfully load it the same way.

> You can change the scanner to allow to specify a test file (better as it doesn't rely on the production file), InformationDisclosureDebugErrorsUnitTest does that. If you prefer to use the same file you can copy it into the (test) ZAP home (given by Constant.getZapHome()).

I'm simply trying to write a unit test that tests the scanner, not rewrite the scanner itself. How, given the current implementation of the scanner, do you get the scanner to successfully load its configuration file during unit test? That's what I'm asking. Simply copying the suspicious-comments.txt file into java/test/zapHomeFiles/xml/ to match the structure of where the same file is in java/main/zapHomeFiles/xml doesn't work.

When I print out the absolute path of zapHomeFiles from a test case, it says it is at something like: C:\Users\Jonathon\AppData\Local\Temp\junit1360484912381141682\home\ but the file suspicious-comments.txt is NOT at \xml\suspicious-comments.txt under this home directory, even when I put the file in pscanrulesBeta\src\test\zapHomeFiles\xml\suspicious-comments.txt. No clue how  I put a file somewhere in the src/test/zapHomeFiles portion of the pscanrulesBeta project so it ends up under junit1360484912381141682\ somewhere in a manner consistent with when ZAP is actually running so the InformationDisclosureSuspiciousComments scanner will successful load the suspicious-comments.txt file during a junit test.

Reply:
> How would a user change that file in the current implementation when using ZAP (before this suggested change)?

The user just has to edit it like any normal text file (in the next run ZAP will load those changes), with the pull request the user would have to edit the file in the add-on (possible but not as simple).

> I'm simply trying to write a unit test that tests the scanner, not rewrite the scanner itself.

It would not require rewrite the scanner (it might be less changes than what was done to load the file from the add-on).

> How, given the current implementation of the scanner, do you get the scanner to successfully load its configuration file during unit test?

As mentioned in the previous comment you need to copy it into the ZAP home, the test infrastructure does not do that. The following changes in the test class are enough:
```Java
@Override
public void setUpZap() throws Exception {
    super.setUpZap();

    Path xmlDir = Files.createDirectories(Paths.get(Constant.getZapHome(), "xml"));
    Path testFile = xmlDir.resolve("suspicious-comments.txt");
    Files.write(testFile, Arrays.asList("# DO", "  FixMe  ", "\t "));
}
```
(Using hardcoded strings, to not depend on the actual file.)

Reply:
OK, I can work with my son to try to make this change but before I do I have this question: In the original change, the suspicious-comments.txt file wasn't moved, it was just loaded in a different way. Does this new way of loading break the ability to edit the file, restart ZAP, and pick up the change? If so, can you explain why?

Reply:
Right, it was not moved but the file being loaded is not the same, the files in `zapHomeFiles` [1] are automatically copied to the ZAP home dir when the add-on is installed and that's the actual file being loaded/used by the scanner. If you change to load the one in the add-on changes done to the file in the home dir will have no effect.

[1] https://github.com/zaproxy/zap-extensions/wiki/AddOnStructure

Reply:
The proposed change loads the file this way: `URL suspiciousCommentsFile = getClass().getResource(suspiciousCommentsListFile);`

Are you sure that loading it this way won't work? i.e., it won't find the file relative to the ZAP home dir in the same way the original code does: `File f = new File(Constant.getZapHome() + File.separator + suspiciousCommentsListFile);`

Reply:
Yes, it won't, the first loads the file in the add-on and the second the file in the file system, for example, if the ZAP home is `/tmp/.ZAP/` the files loaded are:
```
jar:file:/tmp/.ZAP/plugin/pscanrulesBeta-beta-20.zap!/xml/suspicious-comments.txt
/tmp/.ZAP/xml/suspicious-comments.txt
```